### PR TITLE
Make the queue and routing definition API idempotent

### LINKED ISSUE #182: PUT /queue/{queue_name} waited for job completion, if the queue already exists and any job is executing
[PUT /queue/{queue_name}](https://github.com/fireworq/fireworq/blob/master/doc/api.md#api-put-queue) waited for job completion, if the queue already exists and any job is executing.

How to solve

Check if queue exists in advance with [GET /queue/{queue_name}](https://github.com/fireworq/fireworq/blob/master/doc/api.md#api-get-queue)

- check Status 404 or 200
- and is your (will POST) request Body a subset of the body in the response (of GET)

The advantage of idempotent operations is lost because pre-checks need to be performed.

### PR DESCRIPTION
**Pull request by tarao**
Resolve #182

### CODE REVIEW

File: repository/mysql/queue.go [hunk 1]
```diff
@@ -38,12 +44,19 @@ func (r *queueRepository) Add(q *model.Queue) error {
...
+		return updated, err
+	}
+	i, err = res.RowsAffected()
+	if err == nil {
+		updated = updated || (i != 0)
 	}
 
-	return r.updateRevision()
+	if updated {
+		return updated, r.updateRevision()
```
Comment (tarao) [inline repository/mysql/queue.go:57]: When `updateRevision()` is called, it is [captured by `configWatcher`](https://github.com/fireworq/fireworq/blob/66433271014e31587136956ee3535a1bc6ca40a8/service/config_watcher.go#L42), which is [initialized by a `Service`](https://github.com/fireworq/fireworq/blob/66433271014e31587136956ee3535a1bc6ca40a8/service/service.go#L45-L52).

File: repository/inmemory/routing.go [hunk 2]
```diff
@@ -22,12 +24,15 @@ func NewRoutingRepository() repository.RoutingRepository {
...
+func (r *routingRepository) Add(jobCategory string, queueName string) (bool, error) {
 	rs.Lock()
 	defer rs.Unlock()
 
-	rs.m[jobCategory] = queueName
-	return nil
+	if rs.m[jobCategory] != queueName {
+		rs.m[jobCategory] = queueName
+		return true, nil
+	}
```
Comment (yigarashi-9) [inline repository/inmemory/routing.go:34]: [QUESTION] Should we call `updateRevision` here? It is the same for `DeleteByJobCategory`.

**Review by yigarashi-9 — COMMENTED** [review]
I have a question! Please check it out.

File: repository/inmemory/routing.go (hunk 2 above)
Comment (tarao) [reply repository/inmemory/routing.go:35]: Right!  And the reason why the test was passing is a typo [here](https://github.com/fireworq/fireworq/pull/192/commits/b76fedd2732849ac4571b276b6ff22de73287af6#diff-c3b81b4ef722275c696aa684cacaddf9b115a7456ebe95cc58e080b475ba062bL259).  Fixed in [b76fedd](https://github.com/fireworq/fireworq/pull/192/commits/b76fedd2732849ac4571b276b6ff22de73287af6).

**Comment by yigarashi-9** [comment]
LGTM!

**Pull request merged**